### PR TITLE
Fixes #2 for TCP connections: Get the original destination port

### LINKED ISSUE #2: Get the original destination port
Investigate if we get enough information from `/proc/net/nf_conntrack` or `/proc/net/ip_conntrack`

### PR DESCRIPTION
**Pull request by MohammadBilalArif**

### CODE REVIEW

File: glutton.go [hunk 1]
```diff
@@ -20,3 +21,24 @@ func CheckError(err error) {
...
+//return Destination port
+func GetDesport(packetInfo []string, channel *nbc.NonBlockingChan) int {
+	//Receiving conntrack logs from channel
+	stream, ok := <-channel.Recv
+	var connInfo []string
+	for ok {
+		connInfo = stream.([]string)
+
+		if connInfo[2] == packetInfo[0] && connInfo[4] == packetInfo[1] {
+			dport, _ := strconv.Atoi(connInfo[5])
```
Comment (glaslos) [inline glutton.go]: Handle the error

Comment (MohammadBilalArif) [inline glutton.go]: Handled.

File: glutton.go [hunk 2]
```diff
@@ -20,3 +21,24 @@ func CheckError(err error) {
...
 	}
 }
+
+//return Destination port
+func GetDesport(packetInfo []string, channel *nbc.NonBlockingChan) int {
+	//Receiving conntrack logs from channel
+	stream, ok := <-channel.Recv
+	var connInfo []string
+	for ok {
+		connInfo = stream.([]string)
```
Comment (glaslos) [inline glutton.go]: Why do you have to do this?

File: glutton/conntrack.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"github.com/MohammadBilalArif/glutton"
```
Comment (glaslos) [inline glutton/conntrack.go]: This needs to be `mushorg`

File: glutton/conntrack.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+	"os/exec"
+	"regexp"
+	"strconv"
+)
+
+const regexpression = `\[(\d+\.\d+)(?:\s+)?\]\s+\[\w+]\s+\w+\s+\w+\s\w+\s+.+?src=(\d+\.\d+\.\d+\.\d+)\s+dst=(\d+\.\d+\.\d+\.\d+)\s+sport=(\d+)\s+dport=(\d+)\s+`
+
+const connBufferSize int = 30000000
+
+func Monitor_Connections(proto string, channel *nbc.NonBlockingChan) {
```
Comment (glaslos) [inline glutton/conntrack.go]: This is not how you name things in Go

File: glutton/conntrack.go [hunk 5]
```diff
@@ -0,0 +1,58 @@
+package main
```
Comment (glaslos) [inline glutton/conntrack.go]: Why is this in the main package?

File: glutton/conntrack.go [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+	"github.com/hectane/go-nonblockingchan"
+	"log"
+	"os/exec"
+	"regexp"
+	"strconv"
+)
+
+const regexpression = `\[(\d+\.\d+)(?:\s+)?\]\s+\[\w+]\s+\w+\s+\w+\s\w+\s+.+?src=(\d+\.\d+\.\d+\.\d+)\s+dst=(\d+\.\d+\.\d+\.\d+)\s+sport=(\d+)\s+dport=(\d+)\s+`
+
+const connBufferSize int = 30000000
```
Comment (glaslos) [inline glutton/conntrack.go]: Why is this an integer?

File: glutton/conntrack.go [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+func Monitor_Connections(proto string, channel *nbc.NonBlockingChan) {
+	args := []string{
+		"--buffer-size", strconv.Itoa(connBufferSize),
```
Comment (glaslos) [inline glutton/conntrack.go]: What if `strconv.Itoa()` fails?

File: glutton/glutton_server.go [hunk 8]
```diff
@@ -2,32 +2,43 @@ package main
 
 import (
 	"fmt"
+	"github.com/MohammadBilalArif/glutton"
```
Comment (glaslos) [inline glutton/glutton_server.go]: Needs to be mushorg

File: glutton/glutton_server.go [hunk 9]
```diff
@@ -2,32 +2,43 @@ package main
...
-
-	"github.com/mushorg/glutton"
 )
 
-func handleTCPClient(conn net.Conn, filePointer *os.File) {
+func handleTCPClient(conn net.Conn, filePointer *os.File, channel *nbc.NonBlockingChan) {
 	log.SetOutput(filePointer)
+
+	//Spliting addresss to compare with conntrack logs
+	packetInfo := strings.Split(conn.RemoteAddr().String(), ":")
```
Comment (glaslos) [inline glutton/glutton_server.go]: What if `.String()` fails?

File: glutton/glutton_server.go [hunk 10]
```diff
@@ -2,32 +2,43 @@ package main
...
 	"syscall"
-
-	"github.com/mushorg/glutton"
 )
 
-func handleTCPClient(conn net.Conn, filePointer *os.File) {
+func handleTCPClient(conn net.Conn, filePointer *os.File, channel *nbc.NonBlockingChan) {
 	log.SetOutput(filePointer)
+
+	//Spliting addresss to compare with conntrack logs
```
Comment (glaslos) [inline glutton/glutton_server.go]: Spelling

File: glutton/glutton_server.go [hunk 11]
```diff
@@ -2,32 +2,43 @@ package main
...
+	packetInfo := strings.Split(conn.RemoteAddr().String(), ":")
+	dport := glutton.GetDesport(packetInfo, channel)
+
+	if dport == -1 {
+		log.Printf("Packet Droped! [TCP] [ %v ]", conn.RemoteAddr())
```
Comment (glaslos) [inline glutton/glutton_server.go]: Spelling

File: glutton.go (hunk 2 above)
Comment (MohammadBilalArif) [inline glutton.go]: https://github.com/hectane/go-nonblockingchan/issues/1

File: glutton/glutton_server.go [hunk 12]
```diff
@@ -42,9 +53,8 @@ func tcpListener(filePointer *os.File) {
 		if err != nil {
 			continue
 		}
-
-		//goroutines to handle multiple connections
-		go handleTCPClient(tcpConn, filePointer)
+		// //goroutines to handle multiple connections
```
Comment (glaslos) [inline glutton/glutton_server.go]: `// //`

File: glutton/glutton_server.go [hunk 13]
```diff
@@ -83,6 +93,20 @@ func main() {
...
+	//channel for tcp logging
+	tcpChannel := nbc.New()
+	//channel for udp logging
+	udpChannel := nbc.New()
+
+	fmt.Println("Initializaing TCP connections tracking...")
+	go Monitor_Connections("tcp", tcpChannel)
+
+	fmt.Println("Initializaing UDP connections tracking...")
+	go Monitor_Connections("udp", tcpChannel)
```
Comment (glaslos) [inline glutton/glutton_server.go]: tcpChannel?

File: glutton/glutton_server.go [hunk 14]
```diff
@@ -83,6 +93,20 @@ func main() {
...
 	defer filePointer.Close()
 
-	go tcpListener(filePointer)
-	udpListener(filePointer)
+	//channel for tcp logging
+	tcpChannel := nbc.New()
+	//channel for udp logging
+	udpChannel := nbc.New()
+
+	fmt.Println("Initializaing TCP connections tracking...")
```
Comment (glaslos) [inline glutton/glutton_server.go]: spelling

File: glutton.go (hunk 2 above)
Comment (glaslos) [inline glutton.go]: Ah, I see. Make sure you handle assertion errors.

**Comment by glaslos** [comment]
Can you please add some tests?

File: glutton/conntrack.go (hunk 6 above)
Comment (MohammadBilalArif) [inline glutton/conntrack.go]: To increase readability. Well if you do not recommend, I will change.

File: glutton/conntrack.go (hunk 7 above)
Comment (MohammadBilalArif) [inline glutton/conntrack.go]: strconv using hard coded value. I though there is not need to handle. I should edit this.

File: glutton/glutton_server.go (hunk 13 above)
Comment (MohammadBilalArif) [inline glutton/glutton_server.go]: tcpChannel (non-blocking channel) used by the conntrack for tcp logging.

File: glutton.go (hunk 2 above)
Comment (MohammadBilalArif) [inline glutton.go]: OK

File: glutton/conntrack.go (hunk 5 above)
Comment (MohammadBilalArif) [inline glutton/conntrack.go]: I would be grateful if you can help in understanding How I should organize my code? what should be in main and glutton package?

File: glutton/glutton_server.go (hunk 9 above)
Comment (MohammadBilalArif) [inline glutton/glutton_server.go]: I will handle this.

File: glutton/glutton_server.go (hunk 13 above)
Comment (glaslos) [inline glutton/glutton_server.go]: What about udp?

File: glutton/conntrack.go (hunk 5 above)
Comment (glaslos) [inline glutton/conntrack.go]: I try to only have running things in the main package. All functions should go into the glutton package.

Comment (MohammadBilalArif) [inline glutton/conntrack.go]: OK.

File: glutton/glutton_server.go (hunk 13 above)
Comment (MohammadBilalArif) [inline glutton/glutton_server.go]: ucpChannel (non-blocking channel) used by the conntrack for udp logging.

Comment (glaslos) [inline glutton/glutton_server.go]: Spot the error...

Comment (MohammadBilalArif) [inline glutton/glutton_server.go]: Yes.. I am working on udp. I make this PR for showing how I am handling tcp. 
I will send PR soon after solving udp and all the errors you highlighted here.

Comment (glaslos) [inline glutton/glutton_server.go]: Ah I see :) Just add a `# TODO: Implement UPD` the next time.

Comment (MohammadBilalArif) [inline glutton/glutton_server.go]: OK :)

File: glutton/conntrack.go (hunk 4 above)
Comment (MohammadBilalArif) [inline glutton/conntrack.go]: Yup, I will change this.  What about local variable and parameters. I have searched about naming conventions of Go Lang. I could follow [https://talks.golang.org/2014/names.slide#7](url) ?

Comment (glaslos) [inline glutton/conntrack.go]: This is the recommended way: https://golang.org/doc/effective_go.html#names
I use Atom as editor which has a linter which complains if you mess up your names.

**Pull request closed without merging**